python/GafferTest/ReferenceTest.py

```diff
@@ -1150,6 +1150,38 @@ def assertReferenceConnections() :
 
 		assertReferenceConnections()
 
+	def testSearchPaths( self ) :
+
+		s = Gaffer.ScriptNode()
+
+		referenceFile = "test.grf"
+
+		boxA = Gaffer.Box("BoxA")
+		s.addChild(boxA)
+		boxPathA = os.path.join(self.temporaryDirectory(), "a")
+		os.makedirs(boxPathA)
+		fileA = os.path.join(boxPathA, referenceFile)
+		boxA.exportForReference(fileA)
+
+		boxB = Gaffer.Box("BoxB")
+		s.addChild(boxB)
+		boxPathB = os.path.join(self.temporaryDirectory(), "b")
+		os.makedirs(boxPathB)
+		fileB = os.path.join(boxPathB, referenceFile)
+		boxB.exportForReference(fileB)
+
+		s["r"] = Gaffer.Reference()
+
+		searchPathA = ":".join([boxPathA, boxPathB])
+		os.environ["GAFFER_REFERENCE_PATHS"] = searchPathA
+		s["r"].load(referenceFile)
+		self.assertEqual(s["r"].fileName(), fileA)
```

Comment:
Don't we want `fileName() == "test.grf"`, so that when we serialise the script and reload, we re-resolve against the current search path? Basically, it looks like we're currently baking in the resolved filename the first time we load. Sorry, I should have caught this in my original review.

Reply:
You are right - I think the name needs to be resolved in [Reference::fileName()](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/Reference.cpp#L169) and at [this point](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/Reference.cpp#L241) in `Reference::loadInternal()` so `executeFile()` gets the resolved path but the arg is stored as is in `m_fileName`.

I'm assuming the rest of the codebase plays nice and doesn't read raw `m_fileName` so resolving in `Reference::fileName()` should work.

Sound good?

Reply:
> I think the name needs to be resolved in Reference::fileName()

My expectation would be that `Reference::fileName()` returns the unresolved filename, so it always returns the same string that was passed to `Reference::load()` originally. The [Reference serialiser](https://github.com/GafferHQ/gaffer/blob/master/src/GafferModule/SubGraphBinding.cpp#L225-L241) uses `fileName()` to insert an appropriate `load()` call into the .gfr file, and we don't want to serialise the resolved filename at that point. We should add a test case that ensures that we can save/load scripts containing references appropriately without baking the resolution.

There might be a use case for a public method to get at the resolved filename too, but for now I'd be happy with just the unresolved one.

> and at this point in Reference::loadInternal() so executeFile() gets the resolved path

Yep - this seems like the place to do it.

Reply:
Understood - I thought other code may call `fileName()` and expect a resolved path.

Thanks for clarifying.

Reply:
Don't suppose you've had the chance to look into this at all? I'd love to get this merged for the next release...

Reply:
Sure - just pushed a 3rd commit.

I used the same commit message as the first as I'm assuming you will squash-on-merge to make it 1 commit.

Is that the workflow you use? - add commits to address feedback in the PR and then squash-on-merge to keep history neat on the master branch? Thanks